Add Add MapSet.map_set?/1 function

amorphid opened the pull request:
I wanted a function to check if an object is a map set, so I added one!

    iex(1)> MapSet.new() |> MapSet.map_set?()
    true

## Should this edge case a bug?

While adding both ExUnit and DocTest tests, I did bump into an edge case.

    # it's likely no one intentionally builds a MapSet w/ non-true values
    # but I did think of it as a use case
    iex(9)> maybe_map_set = %{__struct__: MapSet, map: %{a: false}}
    #MapSet<[:a]>

    # it walks and talks like a MapSet
    iex(10)> maybe_map_set |> MapSet.map_set?
    true
    iex(11)> %MapSet{} = maybe_map_set
    #MapSet<[:a]>

    # but this just looks weird!
    iex(12)> maybe_map_set |> Map.values
    [MapSet, %{a: false}]

josevalim commented:
MapSet works on the assumption nobody is messing up with its internal and opaque representation. If you do that, things can go wrong, and we can't guarantee correct functioning. So the examples should not include any check such as `%MapSet{...}` because nobody should ever write something like that.

josevalim commented on lib/elixir/lib/map_set.ex:
```diff
@@ -350,6 +350,29 @@ defmodule MapSet do
+      true
+      iex> MapSet.map_set?(MapSet.new([:a,:b]))
+      true
+      iex> MapSet.map_set?(%{__struct__: MapSet, map: %{a: true, b: true}})
```
We should remove this example.

josevalim commented on lib/elixir/lib/map_set.ex:
```diff
@@ -350,6 +350,29 @@ defmodule MapSet do
+      false
+      iex> MapSet.map_set?(%{})
+      false
+      iex> MapSet.map_set?(%{__struct__: MapSet, map: nil})
```
And this one as well.

josevalim commented on lib/elixir/lib/map_set.ex:
```diff
@@ -350,6 +350,29 @@ defmodule MapSet do
+
+  """
+  @spec map_set?(term) :: boolean
+  def map_set?(%MapSet{map: %{}}), do: true
```
Checking for `%MapSet{}` should be enough.

lexmag commented on lib/elixir/lib/map_set.ex:
```diff
@@ -350,6 +350,29 @@ defmodule MapSet do
+
+      iex> MapSet.map_set?(MapSet.new([]))
+      true
+      iex> MapSet.map_set?(MapSet.new([:a,:b]))
```
We need space after comma.

lexmag commented on lib/elixir/lib/map_set.ex:
```diff
@@ -350,6 +350,25 @@ defmodule MapSet do
+  """
+  @spec map_set?(term) :: boolean
+  def map_set?(%MapSet{map: %{}}), do: true
+  def map_set?(_other),            do: false
```
Please don't align `do:` with spaces.

lexmag commented on lib/elixir/test/elixir/map_set_test.exs:
```diff
@@ -81,6 +81,11 @@ defmodule MapSetTest do
   end
 
+  test "map_set?/1" do
+    assert S.map_set?(S.new())        == true
```
Please don't align `==` with spaces.

amorphid commented:
OK, I believe I made all of the requested changes.  Removed any tests that referenced the internal state of maps, and updated requested formatting changes.

lexmag commented:
Thank you @amorphid.

amorphid commented:
@lexmag my pleasure :)  I love how easy y'all make contributing!

whatyouhide commented:
The team discussed this a bit and we decided to remove this feature before releasing it (the PR was reverted in 8a63cd03aca94d7c1a2859b721ff7c0527a23c39) because the exact same `map_set?(term)` functionality can be achieved through `match?(%MapSet{}, term)`. `match?/2` scales very well to all existing structs (`match?(%URI{}, term)` for example), and usually struct that act as types (such as `Regex`, `MapSet`, `Range`) are opaque in their fields but are documented to be those structs, so this is not a "hack" in any sense.

Thanks a lot anyways @amorphid! 💟

amorphid commented:
Okie dokey.  Thanks!  On a different note, I didn't know about match?/1.  One of those "why hadn't I learned that yet?" moments :)
